src/stores/modelStore.ts

```diff
@@ -0,0 +1,133 @@
+import { api } from '@/scripts/api'
+
+/** (Internal helper) finds a value in a metadata object from any of a list of keys. */
+function _findInMetadata(metadata: any, ...keys: string[]): string | null {
+  for (const key of keys) {
+    if (key in metadata) {
+      return metadata[key]
+    }
+    for (const k in metadata) {
+      if (k.endsWith(key)) {
+        return metadata[k]
+      }
+    }
+  }
+  return null
+}
+
+/** Defines and holds metadata for a model */
+export class ComfyModelDef {
+  /** Proper filename of the model */
```

Comment:
We should define the schema in zod for easier parsing and validation.

Reply:
this isn't a schema or api type, it's just data storage. model metadata doesn't come in a reliable format, at least not currently.

Reply:
unless you think we should build a whole "proper metadata" api on the ComfyUI serverside and then just read it for frontend? would make sense to do too tbh

Reply:
Transformed zod code:

```js
import { z } from 'zod'

const zModelMetadata = z.object({
  'modelspec.title': z.string().optional(),
  title: z.string().optional(),
  display_name: z.string().optional(),
  name: z.string().optional(),
  'modelspec.architecture': z.string().optional(),
  architecture: z.string().optional(),
  'modelspec.author': z.string().optional(),
  author: z.string().optional(),
  'modelspec.description': z.string().optional(),
  description: z.string().optional(),
  'modelspec.resolution': z.string().optional(),
  resolution: z.string().optional(),
  'modelspec.usage_hint': z.string().optional(),
  usage_hint: z.string().optional(),
  'modelspec.trigger_phrase': z.string().optional(),
  trigger_phrase: z.string().optional(),
  'modelspec.tags': z.string().optional(),
  tags: z.string().optional(),
  'modelspec.thumbnail': z.string().optional(),
  thumbnail: z.string().optional(),
  image: z.string().optional(),
  icon: z.string().optional()
}).passthrough()

export const zComfyModelDef = z.object({
  name: z.string(),
  directory: z.string(),
  title: z.string(),
  architecture_id: z.string(),
  author: z.string(),
  resolution: z.string(),
  description: z.string(),
  usage_hint: z.string(),
  trigger_phrase: z.string(),
  tags: z.array(z.string()),
  image: z.string(),
  has_loaded_metadata: z.boolean()
})

export type ComfyModelDef = z.infer<typeof zComfyModelDef>

export function parseComfyModelDef(name: string, directory: string, metadata: z.infer<typeof zModelMetadata>): ComfyModelDef {
  const parsedMetadata = zModelMetadata.parse(metadata)
  
  const title = parsedMetadata['modelspec.title'] || 
                parsedMetadata.title || 
                parsedMetadata.display_name || 
                parsedMetadata.name || 
                name

  const tagsString = parsedMetadata['modelspec.tags'] || parsedMetadata.tags || ''
  const tags = tagsString.split(',').map(tag => tag.trim())

  const image = parsedMetadata['modelspec.thumbnail'] || 
                parsedMetadata.thumbnail || 
                parsedMetadata.image || 
                parsedMetadata.icon || 
                ''

  return zComfyModelDef.parse({
    name,
    directory,
    title,
    architecture_id: parsedMetadata['modelspec.architecture'] || parsedMetadata.architecture || '',
    author: parsedMetadata['modelspec.author'] || parsedMetadata.author || '',
    resolution: parsedMetadata['modelspec.resolution'] || parsedMetadata.resolution || '',
    description: parsedMetadata['modelspec.description'] || parsedMetadata.description || '',
    usage_hint: parsedMetadata['modelspec.usage_hint'] || parsedMetadata.usage_hint || '',
    trigger_phrase: parsedMetadata['modelspec.trigger_phrase'] || parsedMetadata.trigger_phrase || '',
    tags,
    image,
    has_loaded_metadata: true
  })
}
```

Reply:
But why tho

> this isn't a schema or api type, it's just temporary local data storage. model metadata doesn't come in a reliable format, at least not currently.

Reply:
If you don't want to define the interface with zod, you can define it in plain TS interface. The current way to initialize the object is not very flexible. Let's say if later we accept the parsed data from backend directly, you would need to adjust the constructor signature.

So my suggestion here is an interface definition ComfyModel, and a static factory function to create ComfyModel from the data source. This also elimiates the middle state where the class is constructed but before the load function is called. (Is there any reason for the middle state to exist?)

Reply:
`Is there any reason for the middle state to exist?` this is the majority state it will usually be in. Since current backend design means loading metadata for a model is a whole special process per-model, it has to be done case-by-case when requested, attempting to load metadata on all models in advance could be extremely slow (minutes of waiting or worse) on some systems. (At least until we rewrite how the backend handles metadata to have a full on-disk cache and all to avoid this, a la how Swarm handles it).

Retooling this after the backend is updated to have better handling is an inevitability either way.

Reply:
I would prefer we define separate interface for the middle state and the final state, so that we know which state the object is in by TS typechecker instead of developer intuition.

Reply:
splitting object type would introduce significant complexity to practical usage, and not make it any easier to tell. It's not defined by "intuition" it's easily checkable with the boolean. More practically in most cases the difference does entirely matter. The times it matters are cases where you actively want the metadata so you'll check the bool and/or `await o.load()` to ensure that.

Reply:
Separate interface is to avoid this boolean flag check everywhere when you need to use the value. Imagine every time you access any field in the structure you have to check a boolean flag and optionally call the load function. The ensure o.load() will also make tracking when the data is fetched very hard.

Reply:
BTW, I'm also fine if you make this explicitly lazy load, i.e. fetch data on first attempt to access fields.

Reply:
this is an explicit lazy load design, yes.

Reply:
Ok, we can continue with this, and refactor later if necessary. Can you add some tests as requested by the previous comment?